File: neetcode-150/two-pointer/15-3sum.cpp
#include <algorithm>
#include <iostream>
#include <set>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
   public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();

        set<tuple<int, int, int>> triplets;
        sort(nums.begin(), nums.end());

        for (int i = 0; i < n - 2; ++i) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            int target = 0 - nums[i];
            int left = i + 1;
            int right = nums.size() - 1;
            while (left < right) {
                int sum = nums[left] + nums[right];
                if (sum == target) {
                    triplets.insert({nums[i], nums[left], nums[right]});
                    left++;
                    right--;
                } else if (sum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        vector<vector<int>> answers;

        for (auto& triplet : triplets) {
            answers.push_back({get<0>(triplet), get<1>(triplet), get<2>(triplet)});
        }
    }
};

File: neetcode-150/two-pointer/15-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        triplets: set[tuple[int, int, int]] = set()
        nums.sort()

        for i in range(n - 2):
            target = 0 - nums[i]
            left = i + 1
            right = n - 1
            while (left < right):
                sum = nums[left] + nums[right]
                if target == sum:
                    triplet = (nums[i], nums[left], nums[right])
                    if not triplet in triplets:
                        triplets.add(triplet)
                    left += 1
                    right -= 1
                elif sum < target:
                    left += 1
                else:
                    right -= 1
        return [list(triplet) for triplet in triplets]
